producer.go

```diff
@@ -18,10 +18,11 @@ type Writer interface {
 }
 
 type producer struct {
-	w Writer
+	w           Writer
+	interceptor *ProducerInterceptor
 }
 
-func NewProducer(cfg *ProducerConfig) (Producer, error) {
+func NewProducer(cfg *ProducerConfig, interceptor *ProducerInterceptor) (Producer, error) {
```

Comment:
we can use `interceptor ...ProducerInterceptor` syntax. 
In that way, client can use 
  - NewProducer(producerConfig)
  - NewProducer(producerConfig, interceptor1)
  - NewProducer(producerConfig, interceptor1, interceptor2)